lib-impl/src/main/java/org/ergoplatform/appkit/impl/OutBoxBuilderImpl.scala

```diff
@@ -39,13 +39,33 @@ class OutBoxBuilderImpl(_txB: UnsignedTransactionBuilderImpl) extends OutBoxBuil
     val tokenNumOfDecVal = token.getMintingBoxR6
     _registers ++= Array(tokenNameVal, tokenDescVal, tokenNumOfDecVal)
 
-    // optional registers, but either all of them or none
-    if (token.getMintingBoxR7 != null && token.getMintingBoxR8 != null) {
-      _registers ++= Array(token.getMintingBoxR7, token.getMintingBoxR8)
-      if (token.getMintingBoxR9 != null)
-        _registers += token.getMintingBoxR9
+    if(token.getMintingBoxR7 != null){
+      _registers ++= Array(token.getMintingBoxR7)
+
+      if(token.getMintingBoxR8 != null){
+        _registers ++= Array(token.getMintingBoxR8)
+
+
+        if(token.getMintingBoxR9 != null){
+          _registers ++= Array(token.getMintingBoxR9)
+        }
+
+      } else {
+        if(token.getMintingBoxR9 != null){
+          throw new IllegalArgumentException("R9 cannot exist without R8")
+        }
+      }
+
+    } else {
+      if(token.getMintingBoxR8 != null){
+        throw new IllegalArgumentException("R8 cannot exist without R7")
+      }
+      if(token.getMintingBoxR9 != null){
+        throw new IllegalArgumentException("R9 cannot exist without R7")
+      }
```

Comment:
That is a lot of nesting. It would be easier to add nulls to _registers and then just check finally that _registers don't have nulls. And throw otherwise. This will be easier to cover with tests.

Reply:
This has some logic error which fails tests, please see how I changed implementation to remove nested statements.